teps/0044-composing-tasks-with-tasks.md

```diff
@@ -0,0 +1,166 @@
+  doing the clone multiple times using [from](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-the-from-parameter))
+
+### Goals
+
+- Make it possible to combine Tasks together so that you can run multiple
+  Tasks as "one unit" (see [Requirements](#requirements):
+  - At runtime in a TaskRun
+  - At Pipeline authoring time
+- Add some of [the features we don't have without PipelineResources](https://docs.google.com/document/d/1KpVyWi-etX00J3hIz_9HlbaNNEyuzP6S986Wjhl3ZnA/edit#)
+  to Tekton Pipelines (without requiring use of PipelineResources)
+  - If possible we can provide a migration path from PipelineResources to the solution proposed here, where applicable
+
+### Non-Goals
+
+- Composing Tasks within Tasks at [Task authoring time](https://github.com/tektoncd/community/blob/master/design-principles.md#reusability).
+  We could decide to include this if we have some use cases that need it; for now avoiding
+  this allows us to avoid many layers of nesting (i.e. Task1 uses Task2 uses Task3, etc.)
+  or even worse, recursion (Task 1 uses Task 2 uses Task 1...)
+- Completely replacing PipelineResources: we could decide to solve this by improving PipelineResources,
+  or we could add a new feature via this TEP and still continue to support PipelineResources
+  (since they provide [more functionality than just composition](https://docs.google.com/document/d/1KpVyWi-etX00J3hIz_9HlbaNNEyuzP6S986Wjhl3ZnA/edit#))
+- This was previously a use case we were targetting but we've decided to descope the TEP slightly,
+  though if we end up solving this problem as well, that's a bonus:
+  * A user wants to use a Task from the catalog with a git repo and doesn't want the extra
+    overhead of using a Pipeline, they just want to make a TaskRun,
+    e.g. [@mattmoor's feedback on PipelineResources and the Pipeline beta](https://twitter.com/mattomata/status/1251378751515922432))
+    where he wants to checkout some code and [use the kaniko task](https://github.com/tektoncd/catalog/tree/master/task/kaniko/0.1)
+    without having to fiddle with volumes
+
+### Use Cases (optional)
+
+- A user wants to use catalog Tasks to checkout code, run unit tests and upload results,
+  and does not want to incur the additional overhead (and performance impact) of creating
+  volume based workspaces to share data between them in a Pipeline. e.g. specifically
+  cloning with [git-clone](https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.2),
+  running tests with [golang-test](https://github.com/tektoncd/catalog/tree/master/task/golang-test/0.1)
+  and uploading results with [gcs-upload](https://github.com/tektoncd/catalog/tree/master/task/gcs-upload/0.1).
+- An organziation does not want to use PVCs at all; for example perhaps they have decided
+  on uploading to and downloading from buckets in the cloud (e.g. GCS)
+- An organization is willing to use PVCs to some extent but needs to put limits on their use
+
+### Overlap with TEP-0046
```

Comment:
@jlpettersson do these look accurate to you?

Reply:
@jlpettersson i think my confusion is the line ["No changes in the Pipeline or Task API for authors"](https://github.com/tektoncd/community/pull/318#discussion_r582316051)

I interpreted this to mean that you'd specify in the PipelineRun that you wanted the entire thing to execute in one pod, i.e. I could write a Pipeline and I could run it as one if I wanted, or I could choose to execute it in multiple pods.

Or maybe you're saying this would be configured at the Controller level, and wouldn't be configured at runtime (in the PipelineRun) OR at authoring time (in the pipeline)?

Reply:
Yes, "configured at the Controller level" is the correct wording for what I meant - at least until this feature is more mature. I'm not against other solutions, but I think it would be good to do that choice later when we see how things goes. My main wish is to solve the concurrency-problem for Tasks in Pipelines. (In my view that problem is the same as affinity-assistant-problem)

So I am not against to have a choice to enable this as a PipelineRun choice in future, but I think it would be good to have it as an experimental thing without API introductions initially.

Reply:
Ah kk, thanks for explaining, that also makes the distinction between the two TEPs pretty clear - I've added some more detail trying to explain why in this TEP I think it's important to be able to provide this configuration at authoring time.